File: frontend/src/apis/customer/UpdateCustomerProfileApi.ts
import type { CustomerId, DeliveryAppState, UpdateCustomerProfileRequest } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath1, defineJsonPostApi1, routeSegment } from '@/system/api/TypedApiDefinitions'

export const updateCustomerProfileApiDefinition = defineJsonPostApi1<CustomerId, UpdateCustomerProfileRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('customers')],
    [routeSegment('profile')],
  )

export function updateCustomerProfile(
  customerId: CustomerId,
  payload: UpdateCustomerProfileRequest,
) {
  return postNormalizedDeliveryState(buildApiPath1(updateCustomerProfileApiDefinition, customerId), payload)
}

File: backend/src/main/scala/services/customer/api/UpdateCustomerProfileApi.scala
package services.customer.api

import domain.shared.given

import cats.effect.IO
import services.customer.utils.updateCustomerProfile
import domain.customer.UpdateCustomerProfileRequest
import domain.shared.{CustomerId, DeliveryAppState, UserRole}
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import system.api.*
import system.app.*

val updateCustomerProfileApi: FixedMethodApi1[CustomerId, DeliveryAppState] =
  jsonPostApi1[CustomerId, UpdateCustomerProfileRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("customers")),
    List(routeSegment("profile")),
  )
